ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
+type IngressNodeFirewallICMPRule struct {
+	// ICMPType define ICMP Type Numbers (RFC 792).
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum:=255
+	// +kubebuilder:validation:Minimum:=0
+	ICMPType uint8 `json:"icmpType,omitempty"`
+
+	// ICMPCode define ICMP Code ID (RFC 792).
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum:=255
+	// +kubebuilder:validation:Minimum:=0
+	ICMPCode uint8 `json:"icmpCode,omitempty"`
+}
+
+// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
+type IngressNodeFirewallProtoRule struct {
+	// Ports can be a single port or range of ports using start-end range format
+	// if admin wanted to filter single port then setting a single port as an integer for example Ports: 80
+	// if admin wanted to filter range of ports then using "start-end" range string format for example Ports: "80-100"
+	//+kubebuilder:validation:Optional
+	Ports intstr.IntOrString `json:"ports,omitempty"`
+}
+
+// IngressNodeProtocolConfig is a discriminated union of protocol's specific configuration.
+// +union
+type IngressNodeProtocolConfig struct {
```

Comment:
You should add a discriminator to this that is an enum of the valid types, then for each type, you need to add a matching field, so 

```
config:
  type: TCP
  tcp:
    ports: 80
```

Reply:
ok fixed that so it looks like this
```yaml
      protocolConfig:
        protocol: TCP
        protoRule:
          ports: "100-200"
```

Reply:
The union members must match the protocol value, you should have one entry for each discriminant assuming you need config for that discriminant
```
 protocolConfig:
        protocol: TCP
        tcp:
          ports: "100-200"
```

Looking at this, would ports better be a list of int or string?
Wondering if this would be more intuitive and would allow individual ports as well as a range 
```
ports:
  - 80
  - 8080
  - 100-200
```